addons/web_editor/models/html_field_history_mixin.py

```diff
@@ -0,0 +1,170 @@
+
+    def write(self, vals):
+        new_revisions = False
+        versioned_fields = self._get_versioned_fields()
+        vals_contain_versioned_fields = set(vals).intersection(versioned_fields)
+        old_contents = {}
+
+        if vals_contain_versioned_fields:
+            self.ensure_one()
+
+            for field_name in versioned_fields:
+                # For security reason we have to ensure the versioned field is
+                # declared as sanitize=True
+                if (
+                    field_name in vals
+                    and not self.env[self._name]._fields[field_name].sanitize
+                ):
+                    raise ValidationError(
+                        "%s field in model %s is not declared as sanitize=True"
+                        % (field_name, self._name)
+                    )
+
+                old_contents[field_name] = self[field_name]
+
+        # Call super().write before generating the patch to be sure we perform
+        # the diff on sanitized data
+        write_result = super().write(vals)
+
+        if not vals_contain_versioned_fields:
+            return write_result
+
+        html_field_history_revisions = (
+            self.html_field_history_revisions
+            if self.html_field_history_revisions
+            else {}
+        )
+
+        for field_name in versioned_fields:
+            new_content = self[field_name]
+
+            if field_name not in html_field_history_revisions:
+                html_field_history_revisions[field_name] = []
+
+            if new_content and self.id and isinstance(vals[field_name], str):
+                old_content = old_contents[field_name]
+                if isinstance(old_content, str) and new_content != old_content:
+                    new_revisions = True
+                    patch = generate_patch(new_content, old_content)
+                    revision_id = (
+                        (
+                            html_field_history_revisions[field_name][0][
+                                "revision_id"
+                            ]
+                            + 1
+                        )
+                        if html_field_history_revisions[field_name]
+                        else 1
+                    )
+
+                    html_field_history_revisions[field_name].insert(
+                        0,
+                        {
+                            "patch": patch,
+                            "revision_id": revision_id,
+                            "create_date": self.env.cr.now().isoformat(),
+                            "create_uid": self.env.uid,
+                            "create_user_name": self.env.user.name,
+                        },
+                    )
+                    html_field_history_revisions[
+                        field_name
+                    ] = html_field_history_revisions[field_name][
+                        : self._max_html_field_history_revisions
+                    ]
```

Comment:
This takes 5 lines for a single operation that doesn't even seem to add any data so I thought the operation itself was somewhat critical or something, then I realized it was only about trimming the number of revisions to the max length... Personally, I would have added a comment about it but maybe I'm the only one that is triggered by this, in which case I guess that would be overkill.

Reply:
I can definitely add a comment